File: src/main/java/com/intel/distml/api/BigModelWriter.scala
package com.intel.distml.api

import com.intel.distml.api.databus.ServerDataBus
import com.intel.distml.util.{KeyRange, Matrix}

/**
 * Created by yunlong on 4/18/15.
 */
class BigModelWriter extends ModelWriter {

  val MAX_FETCH_SIZE = 1000000;

  def writeModel (model: Model, dataBus: ServerDataBus) {

    import scala.collection.JavaConversions._

    for (matrixName <- model.dataMap.keySet) {
      val m: DMatrix = model.dataMap.get (matrixName)
      if (m.`type` == DMatrix.TYPE_PARAM) {
        val size = m.getRowKeys.size()
        println("total param: " + size);

        m.setLocalCache(null)
        var start = 0L;
        while(start < size-1) {
          val end = Math.min(start + MAX_FETCH_SIZE, size) - 1
          val range = new KeyRange(start, end)
          println("fetch param: " + range);
          val result: Matrix = dataBus.fetchFromServer (matrixName, range)
          if (m.localCache == null) {
            m.setLocalCache (result)
          }
          else {
            println("merge to local cache.");
            m.localCache.mergeMatrix(result)
          }
          println("fetch done.");
          start = end + 1;
        }
        println("fetched param: " + matrixName + ", size=" + m.localCache.getRowKeys.size());
      }
    }
  }
}

File: src/main/java/com/intel/distml/api/BigModelWriter.java
package com.intel.distml.api;

import com.intel.distml.api.DMatrix;
import com.intel.distml.api.Model;
import com.intel.distml.api.ModelWriter;
import com.intel.distml.api.databus.ServerDataBus;
import com.intel.distml.util.KeyRange;
import com.intel.distml.util.Matrix;

import java.util.HashMap;

/**
 * Created by yunlong on 4/29/15.
 */
public class BigModelWriter implements ModelWriter {

    private static final int MAX_FETCH_SIZE = 204800000;   // 200M

    protected int maxFetchRows;

    protected HashMap<String, Integer> fetchBatchSizes;

    public BigModelWriter() {
        this(10);
    }

    public BigModelWriter(int estimatedRowSize) {
        maxFetchRows = MAX_FETCH_SIZE / estimatedRowSize;
        fetchBatchSizes = new HashMap<String, Integer>();
    }

    public void setParamRowSize(String matrixName, int rowSize) {
        fetchBatchSizes.put(matrixName, MAX_FETCH_SIZE/rowSize);
    }

    @Override
    public void writeModel (Model model, ServerDataBus dataBus) {

        for (String matrixName : model.dataMap.keySet()) {

            int batchSize = maxFetchRows;
            if (fetchBatchSizes.containsKey(matrixName)) {
                batchSize = fetchBatchSizes.get(matrixName).intValue();
            }
            DMatrix m = model.dataMap.get (matrixName);
            if (m.hasFlag(DMatrix.FLAG_PARAM)) {
                long size = m.getRowKeys().size();
                System.out.println("total param: " + size);

                m.setLocalCache(null);
                long start = 0L;
                while(start < size-1) {
                    long end = Math.min(start + batchSize, size) - 1;
                    KeyRange range = new KeyRange(start, end);
                    System.out.println("fetch param: " + range);
                    Matrix result = dataBus.fetchFromServer (matrixName, range);
                    if (m.localCache == null) {
                        m.setLocalCache (result);
                    }
                    else {
                        System.out.println("merge to local cache.");
                        m.localCache.mergeMatrix(result);
                    }
                    System.out.println("fetch done: " + m.localCache);
                    start = end + 1;
                }
                System.out.println("fetched param: " + matrixName + ", size=" + m.localCache.getRowKeys().size());
            }
        }
    }
}
